# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

importing libraries

## Data Loading and Inspection

### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [3]:
#loading dataset
df = pd.read_csv(
    r"C:\Users\user\dsc-course0-m8-lab\AviationData.csv",
    encoding="ISO-8859-1",
    low_memory=False
)


leading dataset using pandas library

In [4]:
#inspecting NaNs
df.head()

,Event.Id,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
2,20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
3,20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
4,20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


there are 31 columns 

In [5]:
#getting the overview of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

There are bout  88889 entries in the dataset
there are a total of 31 columns , 26 are object datatype while 55 are float datatype

1. inspecting NaNs

In [6]:
#total values of missing data
df.isnull().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38757
Airport.Name              36185
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1382
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7096
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6384
Publication.Date          13771
dtype: i

In [7]:
#total values of missing values in the dataset
df.isnull().sum().sum()

565032

There are  565032 total number of missing values in the dataset

2. DataTypes

In [8]:
#checking for datatypes
df.dtypes

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

 There are both float datatype and object datatype 

3. Perfoming Summary Statistics

In [9]:
#getting the summary statistics
df.describe()

,Number.of.Engines,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured
count,82805.000000,77488.000000,76379.000000,76956.000000,82977.000000
mean,1.146585,0.647855,0.279881,0.357061,5.325440
std,0.446510,5.485960,1.544084,2.235625,27.913634
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,1.000000
75%,1.000000,0.000000,0.000000,0.000000,2.000000
max,8.000000,349.000000,161.000000,380.000000,699.000000


the above is the summary statistics of the numerical columns, we have mean,total count,standard deviation,minimun value,maximum value,the first quartile,the third quatile and the IQR of all the columns.

## Data Cleaning

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

1. Inspecting The relevant columns


In [10]:
#checking the column name and changing the column names
df.columns
df.columns = (
    df.columns
    .str.strip()
    .str.replace(".", "_", regex=False)
    .str.lower()
)
df.head(2)



,event_id,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,...,purpose_of_flight,air_carrier,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996


I replaced the dots in the column names with _ and convert the letters from upper case to lower case

the column name had . which i changes to _ example Report.Status to Report_Status

In [11]:
df.head(2)

,event_id,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,...,purpose_of_flight,air_carrier,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date
0,20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
1,20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996


relevant_columns = [
    # Time filter
    'Event_Date',

    # Aircraft identification
    'Make',
    'Model',
    'Aircraft_Category',

    # Passenger safety
    'Total_Fatal_Injuries',
    'Total_Serious_Injuries',
    'Total_Minor_Injuries',
    'Total_Uninjured',

    # Destruction severity
    'Aircraft_Damage',

    # Conditions / contributing factors
    'Weather_Condition',
    'Broad_phase_of_flight',
    'Location'
]

irrelevant_columns = [
'Event_Id',
'Investigation_Type',
'Accident_Number',
'Country',
'Latitude',
'Longitude',
'Airport_Code',
'Airport_Name',
'Injury_Severity',
'Registration_Number',
'Amateur_Built',
'Number_of_Engines',
'Engine_Type',
'FAR_Description',
'Schedule',
'Purpose_of_flight',
'Air_carrier',
'Report_Status',
'Publication_Date'
]

2. figure out any reasonable imputations

In [12]:
df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [13]:
df['amateur_built'].unique

<bound method Series.unique of 0        No
1        No
2        No
3        No
4        No
         ..
88884    No
88885    No
88886    No
88887    No
88888    No
Name: amateur_built, Length: 88889, dtype: object>

checking for unique values in amateur_build

In [14]:
# i am imputating only the relevant columns
# Filling  numeric values because we assume if the number of inuries were not captured ptobability is there were 
#no injury of any kind this we cant fill with either mean or median
df[['total_fatal_injuries', 'total_serious_injuries', 'total_minor_injuries']] = df[[ 'total_fatal_injuries', 'total_serious_injuries', 'total_minor_injuries' ]].fillna(0)

im assuming that where there is missing values probably there weere no inurede person , so i cant impute it to either mean or meadian

In [15]:
df.info

<bound method DataFrame.info of              event_id investigation_type accident_number  event_date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              location        country   latitude   longitude airport_code  \
0      MOOSE CREEK, ID  Un

In [16]:
#confirm if it it working
df[ 'total_uninjured'].isnull().sum()

5912

we haev 5912 uninjured people

In [17]:
# Fill categorical missing values with unknown bwcause we cant gause
cols=[
   'weather_condition',
    'broad_phase_of_flight',
    'aircraft_damage',
    'aircraft_category',
    'location'
]

df[cols] = df[cols].fillna('Unknown')

filling the missing value with unkown

In [18]:
df['weather_condition'].unique

<bound method Series.unique of 0            UNK
1            UNK
2            IMC
3            IMC
4            VMC
          ...   
88884    Unknown
88885    Unknown
88886        VMC
88887    Unknown
88888    Unknown
Name: weather_condition, Length: 88889, dtype: object>

checking for weather condition missing value after reeplacing with unknown and we find that now there are no missing values

In [19]:
df["weather_condition"].isnull().sum()

0

In [20]:
# Dropping critical missing values in make and model because we cant gause the make or model of an airoplane
df = df.dropna(subset=['make','model' ])

 Dropping critical missing values in make and model because we cant gause the make or model of an airoplane

In [21]:
df['make']

0                           Stinson
1                             Piper
2                            Cessna
3                          Rockwell
4                            Cessna
                    ...            
88884                         PIPER
88885                      BELLANCA
88886    AMERICAN CHAMPION AIRCRAFT
88887                        CESSNA
88888                         PIPER
Name: make, Length: 88777, dtype: object

In [22]:
df.isnull().sum()

event_id                      0
investigation_type            0
accident_number               0
event_date                    0
location                      0
country                     225
latitude                  54430
longitude                 54439
airport_code              38681
airport_name              36117
injury_severity             979
aircraft_damage               0
aircraft_category             0
registration_number        1339
make                          0
model                         0
amateur_built                99
number_of_engines          6023
engine_type                7025
far_description           56841
schedule                  76245
purpose_of_flight          6138
air_carrier               72168
total_fatal_injuries          0
total_serious_injuries        0
total_minor_injuries          0
total_uninjured            5897
weather_condition             0
broad_phase_of_flight         0
report_status              6338
publication_date          13765
dtype: i

checking through the dataset to see which columns still have many misiing values and i will need to drop them 

filter the dataset

In [23]:
df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [24]:
#ill have to keep only data from 1983 to 2023
df['event_date'] = pd.to_datetime(df['event_date'])
df = df[df['event_date'].dt.year >= 1983]

C:\Users\user\AppData\Local\Temp\ipykernel_9416\3589039807.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['event_date'] = pd.to_datetime(df['event_date'])


ill have to keep only data from 1983 to 2023

In [25]:
#ckecking the remaining values of the event date to see whether there is still year<1983
df['event_date'].unique

<bound method Series.unique of 3600    1983-01-01
3601    1983-01-01
3602    1983-01-01
3603    1983-01-01
3604    1983-01-01
           ...    
88884   2022-12-26
88885   2022-12-26
88886   2022-12-26
88887   2022-12-26
88888   2022-12-29
Name: event_date, Length: 85197, dtype: datetime64[ns]>

ckecking the remaining values of the event date to see whether there is still year<1983

In [26]:
#filtering only the category the client want that is 'Airplane'
df = df[df['aircraft_category'] == 'Airplane']
df['aircraft_category'].unique

<bound method Series.unique of 4149     Airplane
4150     Airplane
4171     Airplane
4285     Airplane
5957     Airplane
           ...   
88869    Airplane
88873    Airplane
88876    Airplane
88877    Airplane
88886    Airplane
Name: aircraft_category, Length: 24421, dtype: object>

filtering only the category the client want that is 'Airplane'

In [27]:
print(df['aircraft_category'].value_counts())

aircraft_category
Airplane    24421
Name: count, dtype: int64


we have a total of 24421 airplane

In [28]:
#filter professional aircraft
df = df[df['amateur_built'] == 'No']
df['amateur_built'].unique

<bound method Series.unique of 4149     No
4150     No
4171     No
4285     No
5957     No
         ..
88869    No
88873    No
88876    No
88877    No
88886    No
Name: amateur_built, Length: 21428, dtype: object>

filter professional aircraft

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

Missing values in injury-related columns were carefully handled to avoid introducing bias.
Fatal, serious, and minor injuries were assumed to be zero when missing, as these are typically reported when present.
However, missing values in the total_uninjured column were not imputed, as absence of data does not imply zero uninjured passengers.
Total passengers were computed as the sum of all injury categories.
Rows with incomplete or zero passenger counts were removed to ensure valid rate calculations.

In [29]:
#estimating the total number of passangers
df['total_passengers'] = df[[
    'total_fatal_injuries',
    'total_serious_injuries',
    'total_minor_injuries',
    'total_uninjured'
]].sum(axis=1, skipna=False)
df = df[df['total_passengers'].notna()]
df = df[df['total_passengers'] > 0]
df['total_passengers'].sum()

181591.0

the total number of passagers is 181591
we know that the total number of passegers represent everyone present in the plain both injured and uninjured

In [30]:
#ckeng to see if we have created the total_passengers column
df.head(2)

,event_id,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,...,air_carrier,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date,total_passengers
4149,20001214X42478,Incident,LAX83IA149B,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,...,NaN,0.0,0.0,0.0,588.0,VMC,Standing,Probable Cause,04-12-2014,588.0
4150,20001214X42478,Incident,LAX83IA149A,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,...,"Singapore Airlines, Ltd.",0.0,0.0,0.0,588.0,VMC,Taxi,Probable Cause,04-12-2014,588.0


2. The likelihood of serious/fatal injury

In [31]:
#creating a key safety metrics
df['fatal_injury_rate'] = df['total_fatal_injuries'] / df['total_passengers']#fatal injury rate
df['serious_injury_rate'] = df['total_serious_injuries'] / df['total_passengers']#serious injury rate
df['severe_injury_rate'] = (
    df['total_fatal_injuries'] + df['total_serious_injuries']
) / df['total_passengers']#combined risk


In [32]:
#check to see if we have the sever injury rate column
df.head(2)

,event_id,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,...,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date,total_passengers,fatal_injury_rate,serious_injury_rate,severe_injury_rate
4149,20001214X42478,Incident,LAX83IA149B,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,...,0.0,588.0,VMC,Standing,Probable Cause,04-12-2014,588.0,0.0,0.0,0.0
4150,20001214X42478,Incident,LAX83IA149A,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,...,0.0,588.0,VMC,Taxi,Probable Cause,04-12-2014,588.0,0.0,0.0,0.0


Severe injury rate combines fatal and serious injuries to measure overall passenger risk.

In [33]:
#fixing rows that might have 0 which will break my calculations
df = df[df['total_passengers'] > 0]

In [34]:
#checking for unique values, if the above code worked
df['total_passengers'].unique

<bound method Series.unique of 4149     588.0
4150     588.0
4285       5.0
5957     289.0
5960      26.0
         ...  
88869      1.0
88873      1.0
88876      1.0
88877      1.0
88886      1.0
Name: total_passengers, Length: 19869, dtype: float64>

**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

In [35]:
#inspecting the columns
df['aircraft_damage'].value_counts(dropna=False)
df['aircraft_damage'].unique
df['aircraft_damage'] = df['aircraft_damage'].str.strip().str.lower()#cleaning values in the column to make them similar
df['aircraft_damage'].unique

<bound method Series.unique of 4149           minor
4150           minor
4285         unknown
5957           minor
5960       destroyed
            ...     
88869    substantial
88873    substantial
88876    substantial
88877    substantial
88886    substantial
Name: aircraft_damage, Length: 19869, dtype: object>

inspecting the columns and cleaning values in the column to make them similar 

# I had already handled missing values and named missing values as  unknown ,cleaned  aircraft column by standardizing text lowercasing and trimming whitespace

2. constructing a derived column

In [36]:
df['was_destroyed'] = (df['aircraft_damage'] == 'destroyed').astype(int)
df['was_destroyed'].unique

<bound method Series.unique of 4149     0
4150     0
4285     0
5957     0
5960     1
        ..
88869    0
88873    0
88876    0
88877    0
88886    0
Name: was_destroyed, Length: 19869, dtype: int32>

creating a column was destroyed to represent 1 if the aircraft was destroyed and 0 if otherwise

In [37]:
#to validate my work
df[['aircraft_damage', 'was_destroyed']].head()

,aircraft_damage,was_destroyed
4149,minor,0
4150,minor,0
4285,unknown,0
5957,minor,0
5960,destroyed,1


In [38]:
#still validating my work to get the sum of destroyed aircrafts and otherwise
df['was_destroyed'].value_counts()

was_destroyed
0    17854
1     2015
Name: count, dtype: int64

### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

In [39]:
#inspecting make
df['make'].value_counts(dropna=False).head(20)

make
CESSNA                4791
PIPER                 2776
Cessna                2007
Piper                 1078
BEECH                 1003
BOEING                 642
Beech                  354
MOONEY                 235
CIRRUS DESIGN CORP     218
AIR TRACTOR INC        217
Boeing                 166
BELLANCA               158
AERONCA                149
MAULE                  144
Mooney                 111
AIRBUS                 110
Air Tractor            109
EMBRAER                103
LUSCOMBE                95
CHAMPION                91
Name: count, dtype: int64

Standardize text by converting to lowercase and stripping whitespace
Filter dataset to retain only manufacturers with at least 50 observations for robust analysis
I had name the missing values as uknown

In [40]:
#cleaning the column
df['make'] = df['make'].str.strip().str.lower()
#normalizing common variation
df['Make'] = df['make'].replace({
    'boeing co': 'boeing',
    'boeing company': 'boeing',
    'airbus industrie': 'airbus',
    'cessna aircraft co': 'cessna',
    'piper aircraft corp': 'piper'
})
df['make'].unique

<bound method Series.unique of 4149                       lockheed
4150                         boeing
4285                   de havilland
5957                        douglas
5960                       lockheed
                    ...            
88869                         piper
88873            cirrus design corp
88876                    swearingen
88877                        cessna
88886    american champion aircraft
Name: make, Length: 19869, dtype: object>

In [41]:
#ckecking distribution and if the names are written in the same order
make_counts = df['Make'].value_counts()
make_counts.head(20)

Make
cessna                6822
piper                 3854
beech                 1357
boeing                 810
mooney                 346
cirrus design corp     219
air tractor inc        219
bellanca               208
maule                  208
air tractor            195
aeronca                192
airbus                 167
champion               151
luscombe               135
grumman                133
cirrus                 131
embraer                127
stinson                126
north american         105
dehavilland             95
Name: count, dtype: int64

In [42]:
#keeping make with a reasonable threeshold of 50
valid_makes = make_counts[make_counts >= 50].index
df = df[df['make'].isin(valid_makes)]
df['Make'].value_counts().head(10)

Make
cessna                6798
piper                 3854
beech                 1357
boeing                 808
mooney                 346
cirrus design corp     219
air tractor inc        219
bellanca               208
maule                  208
air tractor            195
Name: count, dtype: int64

this will keep statistical anlysis meaningful

### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

#inspect the model by ;
missing value(NaN)
messy formatting
very rare models


In [43]:
#lets check the model because I had already dropped the missing values
df["model"].head(10)

4150           747
4285         DHC-6
6760       727-200
6806           C35
7084          180K
7708            99
8585     PA-23-250
10247        R172E
12683          441
14357    PA-23-250
Name: model, dtype: object

Remove or handle missing values (NaNs)
Standardize text (lowercase and strip whitespace)
Inspect frequency distribution of models
Check whether model names are unique across manufacturers (Make)
If not unique, create a combined identifier (Make + Model)

In [44]:
df['model'].isnull().sum()#already dropped the missing values

0

now there are 0 missing values in model

In [45]:
df['model'] = (
    df['model']
    .str.lower()
    .str.strip()
    .str.replace(" ", "", regex=False)
)#changing text to lowercase and strip whitespace

changing text to lowercase and strip whitespace

In [46]:
model_counts = df['model'].value_counts()
model_counts.head(20)#inspecting the count and the outcome

model
172          752
182          295
152          295
pa28         289
172s         268
sr22         237
172n         231
180          211
737          199
a36          176
172m         174
150          170
pa-18-150    170
pa-28-140    155
172p         141
140          117
pa18         113
170b         107
172r         106
pa-28-161    100
Name: count, dtype: int64

inspecting the count and the outcome

In [47]:
#normolizing common patterns
df['model'] = df['model'].replace({
    'pa28': 'pa-28',
    'pa18': 'pa-18'
})

normolizing common patterns

In [48]:
#crate a base model column
df['base_model'] = df['model'].str.extract(r'(^[a-zA-Z\-]*\d+)')

In [49]:
#creating final unique aircraft_type
df['aircraft_type'] = df['make'] + "_" + df['base_model']

In [50]:
#validating my work
df[['make', 'model', 'base_model', 'aircraft_type']].head()

,make,model,base_model,aircraft_type
4150,boeing,747,747,boeing_747
4285,de havilland,dhc-6,dhc-6,de havilland_dhc-6
6760,boeing,727-200,727,boeing_727
6806,beech,c35,c35,beech_c35
7084,cessna,180k,180,cessna_180


In [51]:
#comparing safety by aricraft type
df.groupby('aircraft_type')['severe_injury_rate'].mean().sort_values()

aircraft_type
taylorcraft_l2           0.0
bombardier inc_bd7001    0.0
bombardier inc_bd-100    0.0
boeing_pt13              0.0
piper_pa-19              0.0
                        ... 
socata_rallye235         1.0
beech_3                  1.0
aviat_pittss-2           1.0
cessna_fra150            1.0
cessna_c-210             1.0
Name: severe_injury_rate, Length: 770, dtype: float64

In [52]:
df['aircraft_type'].value_counts().head(10)#finding the most common aircraft

aircraft_type
cessna_172     2007
piper_pa-28     965
cessna_182      768
cessna_150      491
piper_pa-18     429
mooney_m20      337
cessna_180      336
boeing_737      308
cessna_152      297
piper_pa28      261
Name: count, dtype: int64

it is clear that cessna_172   is the most common aircraft

### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

In [53]:
df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date', 'total_passengers', 'fatal_injury_rate',
       'serious_injury_rate', 'severe_injury_rate', 'was_destroyed', 'Make',
       'base_model', 'aircraft_type'],
      dtype='object')

In [54]:
#inspecting columns
columns_to_check = [
    'engine_type',
    'weather_condition',
    'number_of_engines',
    'purpose_of_flight',
    'broad_phase_of_flight'
]

for col in columns_to_check:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False).head(10))


--- engine_type ---
engine_type
Reciprocating      12398
NaN                 2534
Turbo Prop           842
Turbo Fan            565
Turbo Jet             64
Unknown               57
Turbo Shaft            6
Geared Turbofan        1
UNK                    1
Name: count, dtype: int64

--- weather_condition ---
weather_condition
VMC        13730
Unknown     1708
IMC          822
Unk          166
UNK           42
Name: count, dtype: int64

--- number_of_engines ---
number_of_engines
1.0    12750
2.0     2194
NaN     1449
4.0       49
3.0       23
0.0        3
Name: count, dtype: int64

--- purpose_of_flight ---
purpose_of_flight
Personal              9522
Instructional         2344
NaN                   2245
Aerial Application     654
Business               388
Positioning            254
Unknown                248
Aerial Observation     143
Skydiving              132
Other Work Use         110
Name: count, dtype: int64

--- broad_phase_of_flight ---
broad_phase_of_flight
Unknown        14

Standardize text (lowercase, strip whitespace)
Check for inconsistent labels (e.g., abbreviations vs full words)
Identify and possibly group similar categories
Retain missing values where appropriate (not all require imputation)

In [55]:
#standardizing text
text_cols = [
    'engine_type',
    'weather_condition',
    'purpose_of_flight',
    'broad_phase_of_flight'
]

for col in text_cols:
    df[col] = df[col].str.strip().str.lower()

standardizing text

In [56]:
#inspecting unique values
for col in text_cols:
    print(f"\n{col} unique values:")
    print(df[col].unique())


engine_type unique values:
['turbo fan' 'turbo prop' 'reciprocating' 'turbo jet' nan 'unknown'
 'turbo shaft' 'geared turbofan' 'unk']

weather_condition unique values:
['vmc' 'unk' 'imc' 'unknown']

purpose_of_flight unique values:
[nan 'skydiving' 'unknown' 'personal' 'business' 'public aircraft'
 'instructional' 'ferry' 'aerial observation' 'aerial application'
 'positioning' 'air race/show' 'other work use' 'flight test' 'glider tow'
 'banner tow' 'public aircraft - state' 'executive/corporate'
 'air race show' 'public aircraft - federal' 'firefighting'
 'public aircraft - local' 'air drop' 'pubs' 'asho']

broad_phase_of_flight unique values:
['taxi' 'standing' 'climb' 'takeoff' 'landing' 'approach' 'descent'
 'unknown' 'cruise' 'maneuvering' 'go-around' 'other']


inspecting unique values

Cleaning of Categorical Columns

these are the cleaning steps appplied;

Standardized text (lowercase, trimmed whitespace)
Unified inconsistent labels (e.g., “unk” → “unknown”)

engine_type:

Standardized naming (e.g., “turbo fan” → “turbofan”)
Grouped similar engine types

weather_condition:

Standardized to core categories: vmc, imc, unknown

purpose_of_flight:

Fixed typos and inconsistencies (e.g., “asho”, “pubs”)
Merged similar categories (e.g., corporate flights)

broad_phase_of_flight:

Standardized naming (e.g., “go-around” → “go around”)
Ensured consistent phase labels

Missing values were retained where appropriate, as they may carry information.

In [57]:
#cleaning engine_type
df['engine_type'] = (
    df['engine_type']
    .str.strip()
    .str.lower()
    .replace({
        'turbo fan': 'turbofan',
        'turbo jet': 'turbojet',
        'turbo prop': 'turboprop',
        'turbo shaft': 'turboshaft',
        'unk': 'unknown'
    })
)

cleaning engine_type

In [58]:
#cleaning weather condition
df['weather_condition'] = df['weather_condition'].replace({
    'unk': 'unknown'
})

cleaning weather condition

In [59]:
#cleaning purpose of flight
df['purpose_of_flight'] = (
    df['purpose_of_flight']
    .str.strip()
    .str.lower()
    .replace({
        'air race show': 'air race/show',
        'asho': 'air race/show',
        'pubs': 'public aircraft',
        'executive/corporate': 'business',
        'public aircraft - federal': 'public aircraft',
        'public aircraft - state': 'public aircraft',
        'public aircraft - local': 'public aircraft'
    })
)

cleaning purpose of flight

In [60]:
#cleaning broad phase of light
df['broad_phase_of_flight'] = (
    df['broad_phase_of_flight']
    .str.strip()
    .str.lower()
    .replace({
        'go-around': 'go around'
    })
)

cleaning broad phase of light

In [61]:
#validate everything
for col in [
    'engine_type',
    'weather_condition',
    'purpose_of_flight',
    'broad_phase_of_flight'
]:
    print(f"\n{col}:")
    print(df[col].value_counts(dropna=False).head(10))


engine_type:
engine_type
reciprocating      12398
NaN                 2534
turboprop            842
turbofan             565
turbojet              64
unknown               58
turboshaft             6
geared turbofan        1
Name: count, dtype: int64

weather_condition:
weather_condition
vmc        13730
unknown     1916
imc          822
Name: count, dtype: int64

purpose_of_flight:
purpose_of_flight
personal              9522
instructional         2344
NaN                   2245
aerial application     654
business               445
positioning            254
unknown                248
aerial observation     143
skydiving              132
other work use         110
Name: count, dtype: int64

broad_phase_of_flight:
broad_phase_of_flight
unknown        14495
landing         1029
takeoff          324
cruise           166
approach         129
taxi              98
maneuvering       66
go around         61
descent           41
climb             30
Name: count, dtype: int64


validate everything

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

calaculating the percentage of missing values to check which columns are mostly empty and which 
columns are still usable

In [62]:
missing_percentage = df.isna().mean().sort_values(ascending=False) * 100
missing_percentage['total_passengers']

0.0

In [63]:
df[['total_fatal_injuries', 'total_serious_injuries']].describe()#checking if my cleaned data is ok

,total_fatal_injuries,total_serious_injuries
count,16468.000000,16468.000000
mean,0.601470,0.299247
std,6.279813,2.436655
min,0.000000,0.000000
25%,0.000000,0.000000
50%,0.000000,0.000000
75%,0.000000,0.000000
max,295.000000,161.000000


checking if my cleaned data is ok, and yes, my data is still ok

In [64]:
df[df['fatal_injury_rate'] > 0][
    ['total_fatal_injuries', 'total_passengers', 'fatal_injury_rate']
].head(10)

,total_fatal_injuries,total_passengers,fatal_injury_rate
4285,1.0,5.0,0.200000
16837,1.0,135.0,0.007407
24630,1.0,3.0,0.333333
25228,2.0,9.0,0.222222
25232,2.0,9.0,0.222222
25965,1.0,2.0,0.500000
27153,1.0,6.0,0.166667
28501,1.0,5.0,0.200000
33711,1.0,11.0,0.090909
37845,1.0,18.0,0.055556


In [65]:
threshold = 0.5

important_cols = [
    'make', 'model',
    'aircraft_damage',
    'total_fatal_injuries',
    'total_serious_injuries',
    'total_minor_injuries',
    'total_uninjured',
    'engine_type',
    'weather_condition',
    'purpose_of_flight',
    'broad_phase_of_flight'
]

cols_to_drop = [
    col for col in df.columns
    if (df[col].isna().mean() > threshold or col in [
        'airport_code', 'airport_name', 'report_status'
    ])
    and col not in important_cols
]

df = df.drop(columns=cols_to_drop)

dropping 'airport_code', 'airport_name', 'report_status'

In [66]:
df.head(2)

,event_id,investigation_type,accident_number,event_date,location,country,latitude,longitude,injury_severity,aircraft_damage,...,broad_phase_of_flight,publication_date,total_passengers,fatal_injury_rate,serious_injury_rate,severe_injury_rate,was_destroyed,Make,base_model,aircraft_type
4150,20001214X42478,Incident,LAX83IA149A,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,Incident,minor,...,taxi,04-12-2014,588.0,0.0,0.0,0.0,0,boeing,747,boeing_747
4285,20001214X42672,Accident,FTW83LA177,1983-04-02,"MCKINNEY, TX",United States,NaN,NaN,Fatal(1),unknown,...,standing,17-10-2016,5.0,0.2,0.0,0.2,0,de havilland,dhc-6,de havilland_dhc-6


Column Selection and Removal

To improve the quality and relevance of the dataset:

Columns with more than 50% missing values were removed, as they provide limited analytical value.
Administrative and location-specific columns (e.g., airport identifiers, report status) were dropped as they do not contribute to aircraft safety analysis.
Key variables related to injury outcomes, aircraft characteristics, and flight conditions were retained.

This ensured that the dataset remained:

Focused on the client’s objectives
Statistically reliable
Interpretable for downstream analysis

what i saw
Larger aircraft tend to have lower fatal injury rates per passenger, while smaller aircraft often show higher variability and higher risk per passenger in accident scenarios.

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

saving the cleaned_aviation data to csv

In [69]:
df.to_csv(
    r"C:\Users\user\dsc-course0-m8-lab\aviation_cleaned_stage1.csv",
    index=False
)